lib/Analysis/Liveness.cpp

```diff
@@ -0,0 +1,379 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+//
+// Implementation of the liveness analysis.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Analysis/Liveness.h"
+#include "mlir/IR/Block.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/Region.h"
+#include "mlir/IR/Value.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace mlir;
+
+/// Builds and holds block information during the construction phase.
+struct BlockInfoBuilder {
+  using ValueSetT = Liveness::ValueSetT;
+
+  /// Fills the block builder with initial liveness information.
+  void build(Block *block) {
+    // Store block for further processing.
+    this->block = block;
+
+    // Mark all block arguments (phis) as defined.
+    for (BlockArgument *argument : block->getArguments())
+      defValues.insert(argument);
+
+    // Check all result values and whether their uses
+    // are inside this block or not (see outValues).
+    for (Operation &operation : *block)
+      for (Value *result : operation.getResults()) {
+        // Mark as defined
+        defValues.insert(result);
+
+        // Check whether this value will be in the outValues
+        // set (its uses escape this block). Due to the SSA
+        // properties of the program, the uses must occur after
+        // the definition. Therefore, we do not have to check
+        // additional conditions to detect an escaping value.
+        for (OpOperand &use : result->getUses())
+          if (use.getOwner()->getBlock() != block) {
+            outValues.insert(result);
+            break;
+          }
+      }
+
+    // Check all operations for used operands.
+    for (Operation &operation : block->getOperations())
+      for (Value *operand : operation.getOperands())
+        useValues.insert(operand);
+  }
+
+  /// Updates live-in information of the current block.
+  /// To do so it uses the default liveness-computation formula:
+  /// newIn = use union out \ def.
+  /// The methods returns true, if the set has changed (newIn != in),
+  /// false otherwise.
+  bool updateLiveIn() {
+    ValueSetT newIn = useValues;
+    newIn.insert(outValues.begin(), outValues.end());
+    // Subtract defValues
+    for (Value *defValue : defValues)
+      newIn.erase(defValue);
+
+    if (newIn.size() == inValues.size())
```

Comment:
Can you add a comment on why checking the size and not the content is enough? I assume there is an invariant that we will never discover that a value is not in the live-ins after its been added originally and so the set can only monotonically grow?